_posts/2023-03-13-llvm-opaque-pointers.md

```diff
@@ -0,0 +1,85 @@
+* Stores the `%x` parameter into `@x` via `%0`;
+* Returns `%x` to the caller. (This caller is normally `Foo.new`, so it goes unused most of the time.)
+
+We can see that the types of `%self` and `%0` are `%Foo*` and `i32*` respectively. These are _typed pointers_, which LLVM has been using for a long time. In LLVM, typed pointers of different pointee types must be explicitly converted using the `bitcast` LLVM instruction, otherwise the resulting LLVM IR is ill-formed. As more and more LLVM frontends came into existence, it was soon realized that typed pointers do not offer many useful semantics, and instead add an unnecessary layer of complexity over IR generation and analysis.
+
+The _opaque pointer_ was first suggested back in [February 2015](https://lists.llvm.org/pipermail/llvm-dev/2015-February/081822.html), where all pointer types would be represented by a single `ptr` in LLVM IR. Then in September 2022, [LLVM 15 now uses opaque pointers by default](https://releases.llvm.org/15.0.0/docs/OpaquePointers.html#version-support), and typed pointers will be removed shortly afterwards. If we compile the program above again, but this time with a Crystal compiler built using LLVM 15, we can see the opaque pointers in action:
+
+```llvm
+; Function Attrs: uwtable(sync)
+define internal i32 @"*Foo#initialize<Int32>:Int32"(ptr %self, i32 %x) #0 {
+entry:
+  %0 = getelementptr inbounds %Foo, ptr %self, i32 0, i32 1
+  store i32 %x, ptr %0, align 4
+  ret i32 %x
+}
+```
+
+To get to there, a few LLVM instructions have to be built differently depending on whether typed or opaque pointers are used, and the `getelementptr` instruction is one such example. The object type used to be inferred from the given pointer, but now that opaque pointers do not carry that information, the IR generator — our Crystal compiler — needs to supply this object type separately. In this case, the `@x` instance variable and the `#initialize` method both belong to `Foo`, so Crystal knows to pass `Foo` to `getelementptr`. But this instruction is also used in a few dozen other places in the compiler, for which no universal migration is applicable.
+
+Work on the migration to opaque pointers started in October 2022, a month after LLVM 15's release, and after countless segfaults and spec failures, [Crystal now supports LLVM 15 on the master branch](https://github.com/crystal-lang/crystal/pull/13173). As it was a rather huge effort, surely we want the opaque pointers to deliver the performance benefits that LLVM promises. So here are some numbers collected from re-building the compiler itself on an Apple M2, first with an LLVM 14 compiler, then with an LLVM 15 one:
+
+* Non-release build:
+  * Codegen (crystal): 2.65s → 2.84s
+  * Codegen (bc+obj): 5.91s → 5.20s
+  * Codegen (linking): 0.76s → 0.34s
+  * dsymutil: 0.35s → 0.39s
+* Release build:
+  * Codegen (bc+obj): 247.86s → 184.37s
+  * Codegen (linking): 0.45s → 0.33s
+  * dsymutil: 0.63s → 0.52s
```

Comment:
I'm just wondering if a table might work better for this. But it's fine as is.

Reply:
It does not seem there are any styles defined for tables yet. I'm not sure if GFM is supported here either

Reply:
Yeah, we'll have nice tables in the new design. We can leave it as a list then.
GFM is supported, btw.